TorchServe IPEX Blog #2 (#2079)
pytorch/tutorials: intermediate_source/torchserve_with_ipex_2.rst

> =====================================================================

Reviewer: ```suggestion
```

Reviewer: Can we please use the following heading syntax:
```
Title
===== 

Heading 1
-----------
Heading 2
~~~~~~~~~~
Heading 3
++++++++++

Author: Thanks @svekars, could you please tell me the syntax for Heading 4

Grokking PyTorch Intel CPU performance from first principles (Part 2)
=====================================================================

> Authors: Min Jean Cho, Jing Xu, Mark Saroufim

Author: Thanks @svekars for the review. Have done so.

, and optimized kernels on CPU via `Intel® Extension for PyTorch* <https://github.com/intel/intel-extension-for-pytorch>`_
, and apply them to TorchServe showcasing 7.71x throughput speedup for ResNet50, and 2.20x throughput speedup for BERT. 

.. figure:: /_static/img/torchserve-ipex-images-2/1.png
   :width: 100%
   :align: center
   
Throughout this blog, we'll use `Top-down Microarchitecture Analysis (TMA) <https://www.intel.com/content/www/us/en/develop/documentation/vtune-cookbook/top/methodologies/top-down-microarchitecture-analysis-method.html>`_ to profile and show that the Back End Bound (Memory Bound, Core Bound) is often the primary bottleneck for under-optimized or under-tuned deep learning workloads, and we'll demonstrate optimization techniques via Intel® Extension for PyTorch* for improving Back End Bound. We'll also use `Intel® VTune™ Profiler's Instrumentation and Tracing Technology (ITT) <https://github.com/pytorch/pytorch/issues/41001>`_ to profile at finer granularity.

> *****************

Reviewer: We typically don't add a table of contents like this because it's autogenerated on the right hand side under Shortcuts. It is difficult to keep a manual TOC like this up-to-date with any future changes in the content.

Author: I'm fine with either options, but I thought TOC would help when readers read the intro since this tutorial has lots of content. Let me know what you'd prefer !

Table of Contents
*****************
- Prerequisite  
    - Top-down Microarchitecture Analysis Method (TMA)
    - Intel® VTune™ Profiler's Instrumentation and Tracing Technology (ITT)
- TorchServe with Intel® Extension for PyTorch*
    - Leveraging Advanced Launcher Configuration: Memory Allocator 	
        - TCMalloc, JeMalloc, PTMalloc
        - Exercise
    - Intel® Extension for PyTorch*
        - Intel® Extension for PyTorch* Optimizations  
        - Intel® Extension for PyTorch* with TorchServe 
        - Exercise

>         

Reviewer: Where is prerequisites? In the TOC above they are listed as first section.

Author: Do we need to have an explicit heading with "Prerequisites" ? The TOC is more of intended for a glance of content at one shot, rather than a link to the sections.

Reviewer: It would be good to have a prerequisites section and list what the user need to know / have installed before they can complete this tutorial.

Author: OK. Let me try moving around some paragraphs, to heave content under each heading.

   :width: 130%
   :align: center
   
The top level, level-1, metrics collect *Retiring*, *Bad Speculation*, *Front End Bound*, *Back End Bound*. The pipeline of CPU can conceptually be simplified and divided into two: the frontend and the backend. The *frontend* is responsible for fetching the program code and decoding them into low-level hardware operations called micro-ops (uOps). The uOps are then fed to the *backend* in a process called allocation. Once allocated, the backend is responsible for executing the uOp in an available execution unit. A completion of uOp's execution is called *retirement*. In contrast, a *bad speculation* is when speculatively fetched uOps are canceled before retiring such as in the case of mis-predicted branches. Each of these metrics can further be broken down in the subsequent levels to pinpoint the bottleneck.

Throughout the blog, we'll use  `toplev <https://github.com/andikleen/pmu-tools/wiki/toplev-manual>`_, a tool part of `pmu-tools <https://github.com/andikleen/pmu-tools>`_ built on top of `Linux perf <https://man7.org/linux/man-pages/man1/perf.1.html>`_, for TMA.

Tune for the Back End Bound
===========================

> The majority of un-tuned deep learning workloads will be Back End Bound. Resolving Back End bound is often resolving sources of latency causing retirement to take longer than necessary. As shown above, Back End Bound has two submetrics – Core Bound and Memory Bound. 

Author: Let's keep as "Back End Bound" to be consistent with the flowchart.

> Memory Bound stalls have causes related to the memory subsystem. For example, last-level cache (LLC or L3 cache) miss causing access to DRAM. Scaling deep learning models often requires significant compute. And high compute utilization requires that data is available when the execution units need it to execute the uOps. This requires prefetching the data and reusing the data in cache instead of fetching that same data multiple times from main memory which causes execution units to be starved while data is being returned. Throughout this blog, we'll show that a more efficient memory allocator, operator fusion, memory layout format optimization reduce overhead on Memory Bound with better cache locality. 

Author: Let's keep as "significant compute".

Operations like GEMM, convolution, deconvolution are compute-intensive. While operations like pooling, batch normalization, activation functions like ReLU are memory-bound.

*********************************************************************
Intel® VTune™ Profiler's Instrumentation and Tracing Technology (ITT)
*********************************************************************
The ITT APIs of Intel® VTune Profiler is a useful tool to annotate a region of your workload for tracing to profile and visualize at a finer granularity of your annotation – OP/function/sub-function granularity. By annotating at the granularity of your PyTorch model's OPs, Intel® VTune Profiler's ITT enables op-level profiling. Intel® VTune Profiler's ITT has been integrated into `PyTorch Autograd Profiler <https://pytorch.org/tutorials/beginner/introyt/autogradyt_tutorial.html#autograd-profiler>`_. :superscript:`1`

> 1. The feature has to be explicitly enabled by with *torch.autograd.profiler.emit_itt()*.

Author: Have changed to "enabled by `with torch.autograd.profiler.emit_itt()`"

> *********************************************

Author: This is the intended TOC in detail (it's little bit hard to see this as a whole going through the rst file) :
![image](https://user-images.githubusercontent.com/93151422/195467746-407f277e-aa48-4421-bab7-31a4601fbaa4.png)

> *********************************************

Author: I'm not sure if I understand fully, but they are not empty. `TorchServe with Intel® Extension for PyTorch*` is a big section with subsections `Leveraging Advanced Launcher Configuration: Memory Allocator`, `Intel Extension for PyTorch`, etc. Similarly, `Leveraging Advanced Launcher Configuration: Memory Allocator` is not empty, it's a section with subsection `TCMalloc, JeMalloc, PTMalloc`.

> *********************************************

Reviewer: Here is the suggestion on what to add to these sections so we don't have three headings one after another:

TorchServe with Intel® Extension for PyTorch*
----------------------------------------------
Intel® Extension for PyTorch*  is a PyTorch extension that provides performance acceleration for deep learning inference and training on Intel CPU. This section provides an example of how you can leverage this extension with TorchServe 

Leveraging Advanced Launcher Configuration: Memory Allocator
~~~~~~
Move line 72 here. 

TCMalloc, JeMalloc, PTMalloc
^^^^^
Both TCMalloc and JeMalloc use thread-local ....

.. figure:: /_static/img/torchserve-ipex-images-2/9.png
   :width: 100%
   :align: center
   
Each inference call is traced in the timeline graph. The duration of the last model inference decreased from 561.688 ms to 251.287 ms - 2.2x speedup.

.. figure:: /_static/img/torchserve-ipex-images-2/10.png
   :width: 100%
   :align: center
   
The timeline graph can be expanded to see op-level profiling results. The duration of *aten::conv2d* decreased from 16.401 ms to 6.392 ms - 2.6x speedup. 

In this section, we've demonstrated that JeMalloc can give better performance than the default PyTorch memory allocator, PTMalloc, with efficient thread-local caches improving Back End Bound.

> Intel® Extension for PyTorch*

Reviewer: This almost looks like a separate tutorial?

Author: It's part of this tutorial, a short introduction to Intel® Extension for PyTorch*, before jumping into TorchServe with Intel® Extension for PyTorch*

=============================

> `Intel® Extension for PyTorch* <https://github.com/intel/intel-extension-for-pytorch>`_ is a Python package to extend PyTorch with optimizations for extra performance boost on Intel hardware. Example optimizations use AVX-512 Vector Neural Network Instructions (AVX512 VNNI) and Intel® Advanced Matrix Extensions (Intel® AMX).

Reviewer: ```suggestion
`Intel® Extension for PyTorch* <https://github.com/intel/intel-extension-for-pytorch>`__ is a Python package to extend PyTorch with optimizations for extra performance boost on Intel hardware. Example optimizations use AVX-512 Vector Neural Network Instructions (AVX512 VNNI) and Intel® Advanced Matrix Extensions (Intel® AMX).
```

.. code:: python
    
    import torch

    class Model(torch.nn.Module):
        def __init__(self):
            super(Model, self).__init__()
            self.conv = torch.nn.Conv2d(16, 33, 3, stride=2)
            self.relu = torch.nn.ReLU()

        def forward(self, x):
            x = self.conv(x)
            x = self.relu(x)
            return x 

    model = Model()
    model.eval()
    data = torch.rand(20, 16, 50, 100)

    #################### code changes ####################
    import intel_extension_for_pytorch as ipex
    model = ipex.optimize(model)
    ######################################################
    
    print(model)

The model consists of 2 operations, Conv2d and ReLU. By printing the model object, we get the following output. 

.. figure:: /_static/img/torchserve-ipex-images-2/11.png
   :width: 60%
   :align: center
   
Let's collect level-1 TMA metrics. 

.. figure:: /_static/img/torchserve-ipex-images-2/12.png
   :width: 100%
   :align: center
   
Notice the Back End Bound reduced from 68.9 to 38.5 – 1.8x speedup.

Additionally, let's profile with PyTorch Profiler. 

.. figure:: /_static/img/torchserve-ipex-images-2/13.png

>    :width: 150%

Author: All images are already changed to less than 100% width. The comment is on outdated file, prior to this change.

   :align: center
   
Notice the CPU time reduced from 851 us to 310 us – 2.7X speedup. 

Graph Optimization
------------------

> It is highly recommended for users to take advantage of Intel® Extension for PyTorch* with `TorchScript <https://pytorch.org/docs/stable/jit.html>`_ for further graph optimizations. To optimize performance further with TorchScript, Intel® Extension for PyTorch* supports oneDNN fusion of frequently used FP32/BF16 operator patterns, like Conv2D+ReLU, Linear+ReLU, etc to reduce operator/kernel invocation overheads, and for better cache locality. Some operator fusions allow to maintain temporary calculations, data type conversions, data layouts for better cache locality. As well as for INT8, Intel® Extension for PyTorch* has built-in quantization recipes to deliver good statistical accuracy for popular DL workloads including CNN, NLP and recommendation models. The quantized model is then optimized with oneDNN fusion support. 

Reviewer: ```suggestion
It is highly recommended for users to take advantage of Intel® Extension for PyTorch* with `TorchScript <https://pytorch.org/docs/stable/jit.html>`_ for further graph optimizations. To optimize performance further with TorchScript, Intel® Extension for PyTorch* supports oneDNN fusion of frequently used FP32/BF16 operator patterns, like Conv2D+ReLU, Linear+ReLU, or other, to reduce operator/kernel invocation overheads, and for better cache locality. Some operator fusions allow to maintain temporary calculations, data type conversions, data layouts for better cache locality. As well as for INT8, Intel® Extension for PyTorch* has built-in quantization recipes to deliver good statistical accuracy for popular DL workloads including CNN, NLP and recommendation models. The quantized model is then optimized with oneDNN fusion support. 
```

Author: have changed etc to and more.

Exercise
^^^^^^^^ 
Let's profile FP32 graph optimization with TorchScript. 

> As in the previous exercises, we'll bind the workload to physical cores of the first socket.

Reviewer: ```suggestion
As in the previous exercises, we will bind the workload to physical cores of the first socket.
```

   
Notice the Back End Bound reduced from 67.1 to 37.5 – 1.8x speedup.

Additionally, let's profile with PyTorch Profiler. 

.. figure:: /_static/img/torchserve-ipex-images-2/15.png
   :width: 150%
   :align: center
   
Notice that with Intel® Extension for PyTorch*  Conv + ReLU operators are fused, and the CPU time reduced from 803 us to 248 us – 3.2X speedup. The oneDNN eltwise post-op enables fusing a primitive with an elementwise primitive. This is one of the most popular kinds of fusion: an eltwise (typically an activation function such as ReLU) with preceding convolution or inner product. Have a look at the oneDNN verbose log shown in the next section.

Channels Last Memory Format
---------------------------
When invoking *ipex.optimize* on model, Intel® Extension for PyTorch* automatically converts the model to optimized memory format, channels last. Channels last is a memory format that is more friendly to Intel Architecture. Compared to PyTorch default channels first NCHW (batch, channels, height, width) memory format, channels last NHWC (batch, height, width, channels) memory format generally accelerates convolutional neural networks with better cache locality. 

> One thing to note is that it is expensive to convert memory format. So it's better to convert the memory format prior to deployment once, and keep the memory format conversion minimum during deployment. As the data propagates through model's layers the channels last memory format is preserved through consecutive channels last supported layers (e.g., Conv2d -> ReLU -> Conv2d) and conversions are only made in between channels last unsupported layers. Refer to `Memory Format Propagation <https://www.intel.com/content/www/us/en/develop/documentation/onednn-developer-guide-and-reference/top/programming-model/memory-format-propagation.html>`_ for more details.

Reviewer: ```suggestion
One thing to note is that it is expensive to convert memory format. So it's better to convert the memory format prior to deployment once, and keep the memory format conversion minimum during deployment. As the data propagates through model's layers the channels last memory format is preserved through consecutive channels last supported layers (for example, Conv2d -> ReLU -> Conv2d) and conversions are only made in between channels last unsupported layers. See `Memory Format Propagation <https://www.intel.com/content/www/us/en/develop/documentation/onednn-developer-guide-and-reference/top/programming-model/memory-format-propagation.html>`_ for more details.
```

Exercise
^^^^^^^^
Let's demonstrate channels last optimization. 

.. code:: python

    import torch

    class Model(torch.nn.Module):
        def __init__(self):
            super(Model, self).__init__()
            self.conv = torch.nn.Conv2d(16, 33, 3, stride=2)
            self.relu = torch.nn.ReLU()

        def forward(self, x):
            x = self.conv(x)
            x = self.relu(x)
            return x 

    model = Model()
    model.eval()
    data = torch.rand(20, 16, 50, 100)

    import intel_extension_for_pytorch as ipex
    ############################### code changes ###############################
    ipex.disable_auto_channels_last() # omit this line for channels_last (default) 
    ############################################################################
    model = ipex.optimize(model)

    with torch.no_grad():
        model = torch.jit.trace(model, data)
        model = torch.jit.freeze(model)

> We'll use `oneDNN verbose mode <https://oneapi-src.github.io/oneDNN/dev_guide_verbose.html>`_, a tool to help collect information at oneDNN graph level such as operator fusions, kernel execution time spent on executing oneDNN primitives. For more information, refer to the `oneDNN Documentation <https://oneapi-src.github.io/oneDNN/index.html>`_.

Reviewer: ```suggestion
We will use `oneDNN verbose mode <https://oneapi-src.github.io/oneDNN/dev_guide_verbose.html>`_, a tool to help collect information at oneDNN graph level such as operator fusions, kernel execution time spent on executing oneDNN primitives. For more information, refer to the `oneDNN Documentation <https://oneapi-src.github.io/oneDNN/index.html>`_.
```

   
.. figure:: /_static/img/torchserve-ipex-images-2/17.png
   :width: 100%
   :align: center
   
Above is oneDNN verbose from channels first. We can verify that there are reorders from weight and data, then do computation, and finally reorder output back.   

.. figure:: /_static/img/torchserve-ipex-images-2/18.png
   :width: 80%
   :align: center
   
Above is oneDNN verbose from channels last. We can verify that channels last memory format avoids unnecessary reorders.

Intel® Extension for PyTorch* Integration into TorchServe
========================================================= 
Intel® Extension for PyTorch* has already been integrated into TorchServe to improve the performance out-of-box. :superscript:`2` When TorchServe creates single or multiple instances for inference, the launcher from Intel® Extension for PyTorch* is invoked automatically to set CPU affinity to these instances when launcher is enabled. :superscript:`3` This helps to make each instance use its assigned resources as high efficiently as possible, and minimize resource conflict among instances. For custom handler scripts, we recommend adding the *intel_extension_for_pytorch* package in.

> 2. The feature has to be explicitly enabled by setting *ipex_enable=true* in  *config.properties*.

Reviewer: should lines 383 and 384 be converted int one?

Author: Let's keep the two features (and lines 383, 384) separate to be explicit.

3. The feature has to be explicitly enabled by setting *cpu_launcher_enable=true* in  *config.properties*.

Performance Boost with Intel® Extension for PyTorch*
==================================================== 
Below summarizes performance boost of TorchServe with Intel® Extension for PyTorch* for ResNet50 and BERT-base-uncased. 

.. figure:: /_static/img/torchserve-ipex-images-2/19.png
   :width: 100%
   :align: center
   
Exercise with TorchServe
========================
Let's profile Intel® Extension for PyTorch* optimizations with TorchServe. 

> We'll use `TorchServe apache-bench benchmarking <https://github.com/pytorch/serve/tree/master/benchmarks#benchmarking-with-apache-bench>`_ with ResNet50 FP32 TorchScript, batch size 32, concurrency 32, requests 8960. All other parameters are the same as the `default parameters <https://github.com/pytorch/serve/tree/master/benchmarks#benchmark-parameters>`_. 

Reviewer: ```suggestion
We will use `TorchServe apache-bench benchmarking <https://github.com/pytorch/serve/tree/master/benchmarks#benchmarking-with-apache-bench>`_ with ResNet50 FP32 TorchScript, batch size 32, concurrency 32, requests 8960. All other parameters are the same as the `default parameters <https://github.com/pytorch/serve/tree/master/benchmarks#benchmark-parameters>`_. 
```

> As in the previous exercise, we'll use the launcher to bind the workload to physical cores of the first socket. To do so, user simply needs to add a few lines in `config.properties <https://github.com/pytorch/serve/tree/master/benchmarks#benchmark-parameters>`_: 

Reviewer: ```suggestion
As in the previous exercise, we will use the launcher to bind the workload to physical cores of the first socket. To do so, user simply needs to add a few lines in `config.properties <https://github.com/pytorch/serve/tree/master/benchmarks#benchmark-parameters>`_: 
```

.. code:: python

    cpu_launcher_enable=true
    cpu_launcher_args=--node_id 0

Let's collect level-1 TMA metrics. 

.. figure:: /_static/img/torchserve-ipex-images-2/20.png
   :width: 100%
   :align: center
   

> Level-1 TMA shows that both are bounded by the backend. As discussed earlier, the majority of un-tuned deep learning workloads will be Back End Bound. Notice the Back End Bound reduced from 70.0 to 54.1. Let's go one level deeper. 

Reviewer: ```suggestion
Level-1 TMA shows that both are bounded by the backend. As discussed earlier, the majority of untuned deep learning workloads will be Back End Bound. Notice the Back End Bound reduced from 70.0 to 54.1. Let's go one level deeper. 
```

   :width: 100%
   :align: center
   
Scaling deep learning models for production on a model serving framework like TorchServe requires high compute utilization. This requires that data is available through prefetching and reusing the data in cache when the execution units need it to execute the uOps. Level-3 TMA shows that the Back End Memory Bound improved from DRAM Bound to Core Bound. 

As in the previous exercise with TorchServe, let's use Intel® VTune Profiler ITT to annotate `TorchServe inference scope <https://github.com/pytorch/serve/blob/master/ts/torch_handler/base_handler.py#L188>`_ to profile at inference-level granularity.

.. figure:: /_static/img/torchserve-ipex-images-2/23.png
   :width: 100%
   :align: center
   
Each inference call is traced in the timeline graph. The duration of the last inference call decreased from 215.731 ms to 95.634 ms - 2.3x speedup.

.. figure:: /_static/img/torchserve-ipex-images-2/24.png
   :width: 100%
   :align: center
   
The timeline graph can be expanded to see op-level profiling results. Notice that Conv + ReLU has been fused, and the duration decreased from 6.393 ms + 1.731 ms to 3.408 ms - 2.4x speedup. 

**********

> Take-aways

Reviewer: ```suggestion
Conclusion
```

********** 

> In this blog, we've used Top-down Microarchitecture Analysis (TMA) and Intel® VTune™ Profiler's Instrumentation and Tracing Technology (ITT) to demonstrate that 

Reviewer: ```suggestion
In this tutorial, we have used Top-down Microarchitecture Analysis (TMA) and Intel® VTune™ Profiler's Instrumentation and Tracing Technology (ITT) to demonstrate that 
```
